docs/rfc/column-level-lineage.md

```diff
@@ -0,0 +1,726 @@
+
+## Basic example
+Here are sample `json` requests to create dataset `p_foo_13`, `p_2_foo_13` and `c_bar_13`. 
+>For `p_foo_13` and `p_2_foo_13`, it is nothing new. I put them in the later. For `c_bar_13`, we introduced new aspect of `FineGrainUpStreamLineage`.
+1. dataset request Json for c_bar_13
+```
+{
+    "snapshot": {
+        "aspects": [
+            {
+                "com.linkedin.common.Ownership": {
+                    "owners": [
+                        {
+                            "owner": "urn:li:corpuser:kzhang13",
+                            "type": "DATAOWNER"
+                        }
+                    ],
+                    "lastModified": {
+                        "time": 0,
+                        "actor": "urn:li:corpuser:kzhang13"
+                    }
+                }
+            },
+            {
+                "com.linkedin.dataset.UpstreamLineage": {
+                    "upstreams": [
+                        {
+                            "auditStamp": {
+                                "time": 0,
+                                "actor": "urn:li:corpuser:kzhang13"
+                            },
+                            "dataset": "urn:li:dataset:(urn:li:dataPlatform:hasketl,p_foo_13,PROD)",
+                            "type": "TRANSFORMED"
+                        },
+                        {
+                            "auditStamp": {
+                                "time": 0,
+                                "actor": "urn:li:corpuser:kzhang13"
+                            },
+                            "dataset": "urn:li:dataset:(urn:li:dataPlatform:hasketl,2_p_foo_13,PROD)",
+                            "type": "TRANSFORMED"
+                        }
+                    ]
+                }                  
+            },
+            {
+                "com.linkedin.dataset.FineGrainUpstreamLineage": {
+                    "upstreams": [
+                        {
+                            "auditStamp": {
+                                "time": 0,
+                                "actor": "urn:li:corpuser:kzhang13"
+                            },
+                            "dataset": "urn:li:dataset:(urn:li:dataPlatform:hasketl,p_foo_13,PROD)",
+                            "fields": [
+                                {
+                                    "sourceField": "foo131",
+                                    "targetField" : "bar131"
+                                },
+                                {
+                                    "sourceField": "foo132",
+                                    "targetField" : "bar132"
+                                }
+                            ]
```

Comment:
We might have to start thinking field as a first class entity. And, the relations to the source fields can come from multiple datasets. Imagine, two fields from two different datasets can get transformed to form a destination field. 

```
record DatasetFieldMapping includes BaseFieldMapping{
 /**
  * Source fields from which the fine grained lineage is derived
  */
 sourceFields: union[DatasetFieldUrn]

 /**
  * Destination field which is derived from source fields
  */
 destinationField: DatasetFieldUrn
```

Reply:
Right, source fields can be more than one. I define the 
```
record FineGrainUpstream {
  auditStamp: AuditStamp
  dataset: DatasetUrn
  fields: array[FineGrainFieldUpstream]
}
```
The `fields` is an array of the FieldMapping, which, in my opinion, could work with current neo4j implementation, and generate `n:1` relationship between a group of `sourceField` and `a destination`. 
The reason to implement in this way is to get around of the problem that neo4j only recognizes the first class entity. 

For sure, if we have made datasetField as a new entity, the problem above might not be existed. It however will create different problem on the neo4j graph creation.

Reply:
I have thought about the DatasetField as the first class entity . As I remember, `Schema` being a first class entity is on the [road map](https://github.com/linkedin/datahub/blob/master/docs/roadmap.md#schemas-as-an-entity). To me, we should onboard `schema` before onboarding `datasetField`.